src/common/rate_limit/src/lib.rs

```diff
@@ -0,0 +1,600 @@
+        let (tx, rx) = oneshot::channel();
+        self.waiters.lock().push(tx);
+        Check::RetryAfter(rx)
+    }
+}
+
+#[derive(Debug)]
+pub struct FixedRateLimiter {
+    inner: LeakBucket,
+    rate: NonZeroU64,
+}
+
+impl FixedRateLimiter {
+    pub fn new(rate: NonZeroU64) -> Self {
+        let inner = LeakBucket::new(rate);
+        Self { inner, rate }
+    }
+}
+
+impl RateLimiterTrait for FixedRateLimiter {
+    fn rate_limit(&self) -> RateLimit {
+        RateLimit::Fixed(self.rate)
+    }
+
+    fn check(&self, quota: u64) -> Check {
+        match self.inner.check(quota) {
+            Ok(()) => Check::Ok,
+            Err(duration) => Check::Retry(duration),
+        }
+    }
+}
+
+/// A GCRA-like leak bucket visual scheduler that never deny request even whose weight is larger than tau and only count TAT.
+#[derive(Debug)]
+pub struct LeakBucket {
+    /// Weight scale per 1.0 unit quota in nanosecond.
+    ///
+    /// scale is always non-zero.
+    ///
+    /// scale = rate / 1 (in second)
+    scale: AtomicU64,
+
+    /// Last request's TAT (Theoretical Arrival Time) in nanosecond.
+    ltat: AtomicU64,
+
+    /// Zero time instant.
+    origin: Instant,
+
+    /// Total allowed quotas.
+    total_allowed_quotas: AtomicU64,
+    /// Total waited nanos.
+    total_waited_nanos: AtomicI64,
+}
+
+impl LeakBucket {
+    const NANO: u64 = Duration::from_secs(1).as_nanos() as u64;
+
+    /// calculate the weight scale per 1.0 unit quota in nanosecond.
+    fn scale(rate: NonZeroU64) -> u64 {
+        std::cmp::max(Self::NANO / rate.get(), 1)
+    }
+
+    /// Create a new GCRA-like leak bucket visual scheduler with given rate.
+    fn new(rate: NonZeroU64) -> Self {
+        let scale = Self::scale(rate);
+
+        let origin = Instant::now();
+        let scale = AtomicU64::new(scale);
+
+        Self {
+            scale,
+            ltat: AtomicU64::new(0),
+            origin,
+            total_allowed_quotas: AtomicU64::new(0),
+            total_waited_nanos: AtomicI64::new(0),
+        }
+    }
+
+    /// Check if the request with the given quota is supposed to be allowed at the moment.
+    ///
+    /// On success, the quota will be consumed. The caller is supposed to proceed the quota.
+    ///
+    /// On failure, the minimal duration to retry `check()` is returned.
+    fn check(&self, quota: u64) -> Result<(), Duration> {
+        let now = Instant::now();
+        let tnow = now.duration_since(self.origin).as_nanos() as u64;
+
+        let weight = quota * self.scale.load(Ordering::Relaxed);
+
+        let mut ltat = self.ltat.load(Ordering::Acquire);
+        let tat = loop {
+            let tat = ltat + weight;
+
+            if tat > tnow {
+                self.total_waited_nanos
+                    .fetch_add((tat - tnow) as i64, Ordering::Relaxed);
+                return Err(Duration::from_nanos(tat - tnow));
+            }
+
+            let ltat_new = std::cmp::max(tat, tnow);
```

Comment:
When we reach here, `tat <= tnow` must hold. In this case, should we set ltat to `tat` instead of `tnow`? Otherwise, we will consume more quota than the provided one.

Reply:
The `max` here is necessary. I forgot to remove it after adding the early return above.

However, the `ltat` is supposed to be set to `tnow` (the larger one) to prevent form quota accumulation with no limitation.

e.g.

```plain
rate limit: 1 row/s

t(0s) ~ t(100s): no requests
t(101s) ~ t(110s): 2 rows
``` 

If we set the `ltat` to `tat`, during 101s ~ 110s, the 2 rows requests will be allowed with no delay, which makes the avg rate during the 10 seconds larger than expected.

But if we set the `ltat` to `tnow` here,  the request has to be delayed after each 2 row request is allowed, to make the avg rate during the 10 seconds reasonable.

Reply:
I see. I thought we intended to support quota accumulation. Good to know that is not the case. Thanks for the explanation.